### PR TITLE
Klaus/crosshair angular mousetooltip

### PR DESCRIPTION
**Pull request by klues**
This is a new version of https://github.com/asterics/AsTeRICS/pull/281
I created this new branch in order to have a stable version in the other branch for the tests with our client (which unfortunately was'nt realized until yet).

This PR includes:
1) new plugin CrosshairCursorControl
2) new plugin AngularCursorControl
3) new plugin Tooltip -> tooltip functionality that at first was included in CrosshairCursorControl now as separate plugin in order to allow flexible usage
4) demo models for CrosshairCursorControl, all in `bin\ARE\models\useCaseDemos\mouseControl\crosshairCursorControl*.acs`, including test models for keyboard and EMG singals.
5) demo models for AngularCursorControl, all in `bin\ARE\models\useCaseDemos\mouseControl\angularCursorControl*.acs`, including test models for keyboard and EMG singals.
6) new model `C:\data\src\AsTeRICS\bin\ARE\models\XFaceTrackerMouse_tooltips.acs` that integrates the Tooltip plugin into the cameramouse
7) Adaption of the Mouse plugin -> added output ports x/y and event listener `moveToLastStable` for interaction with Tooltip plugin
8) updated/added documentation for CrosshairCursorControl, AngularCursorControl, Tooltip and Mouse plugins

@ChrisVeigl @deinhofer @benjaminaigner 
if you want, please review and/or test the demo models, so that we can merge.

@ point (6), cameramouse with tooltips:
if you want to try this model, after starting:
move your head/mouse around the screen some times and then continue normal, slower moving. this does some sort of calibration at the beginning. Afterwards nodding or shaking the head activates tooltips and nodding up/down or shaking left/right selects the next/previous tooltip.

### CODE REVIEW

File: ARE/components/actuator.tooltip/src/main/java/eu/asterics/component/actuator/tooltip/GUI.java [hunk 1]
```diff
@@ -67,139 +67,161 @@
...
+        showTooltip(false);
+        this.tooltipActive = false;
     }
 
+    /**
+     * Selects the next index of Tooltip images. If the @see {@link #TOOLTIP_INDEX_MAX} is reached, starts with {@link #TOOLTIP_INDEX_MIN}.
+     */
     void navigateNextTooltip() {
-        navigateTooltips(1);
+        actTooltip = (actTooltip + 1 > TOOLTIP_INDEX_MAX) ? TOOLTIP_INDEX_MIN : actTooltip + 1;
```
Comment (klues) [inline ARE/components/actuator.tooltip/src/main/java/eu/asterics/component/actuator/tooltip/GUI.java]: the intended and previously implemented functionality was, that a navigation out of the bounds of existing .png files does deactivate the tooltips which isn't the case anymore now. In this implementation the user cannot leave the tooltip mode without providing an extra command/event to deactivate tooltips, but users should be able to deactivate tooltips by the same actions that they are doing for navigating through the tooltips.

**Comment by deinhofer** [comment]
> thanks for the refactoring, however unfortunately many things do not work as intended anymore (as far as I see from code), please refine.

Sorry for the heavy refactoring, but the code was really very buggy and bad style. So I refactored many parts. Let's reenable missing functionality together, it shouldn't be a big issue now.

File: ARE/components/actuator.tooltip/src/main/java/eu/asterics/component/actuator/tooltip/GUI.java [hunk 2]
```diff
@@ -67,139 +67,161 @@
...
-            g.drawImage(image, toolX, toolY, null);
-        }
+        /*
+         * if (image != null) { int toolX, toolY; if ((int) mouseY < image.getHeight()) { toolY = (int) mouseY + 10; } else { toolY = (int) mouseY -
+         * image.getHeight() - 10; } if ((int) mouseX < image.getWidth()) { toolX = (int) mouseX + 10; } else { toolX = (int) mouseX - image.getWidth() - 10; }
+         * 
+         * g.drawImage(image, toolX, toolY, null); }
+         */
+        // System.out.println("Drawing image: "+image+", retVal: "+g.drawImage(image, 0, 0, null));
+        g.drawImage(image, 0, 0, null);
```
[the remark this answers is not in the document]
Comment (deinhofer) [reply ARE/components/actuator.tooltip/src/main/java/eu/asterics/component/actuator/tooltip/GUI.java]: I was not sure if this is really needed. I first wanted to check if it works generally.
The problem with the old drawing was, that it created a screensized transparent JFrame and positioned the image inside. Now the JFrame is just of image size and the JFrame is positioned to the respective location.

File: ARE/components/actuator.tooltip/src/main/java/eu/asterics/component/actuator/tooltip/TooltipInstance.java [hunk 3]
```diff
@@ -218,87 +228,91 @@ public Object setRuntimePropertyValue(String propertyName, Object newValue) {
...
     };
 
     final IRuntimeEventListenerPort elpPreviousTooltip = new IRuntimeEventListenerPort() {
         public void receiveEvent(final String data) {
-            if (gui.tooltipsActive()) {
-                lastActionTime = System.currentTimeMillis();
-                gui.navigatePreviousTooltip();
-                if (!gui.tooltipsActive()) etpTooltipDeactivated.raiseEvent();
-            }
+            // cancelSelectionTimer(true);
```
[the remark this answers is not in the document]
Comment (deinhofer) [reply ARE/components/actuator.tooltip/src/main/java/eu/asterics/component/actuator/tooltip/TooltipInstance.java]: ok, that's what I wanted to discuss with also.

File: ARE/components/actuator.tooltip/src/main/java/eu/asterics/component/actuator/tooltip/GUI.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (klues) [reply ARE/components/actuator.tooltip/src/main/java/eu/asterics/component/actuator/tooltip/GUI.java]: yeah, it's of course better to have an image-sized transparent JFrame. I think a smiliar logic as before now belongs to the code part where you set the location of the JFrame.

**Comment by klues** [comment]
No problem, thats what a review is for ;)
The "problem" is that we need this plugin today for our evaluation with the client, but I will simply use an earlier commit.

answering your remaining questions:
> Are the properties propInitX and propInitY really necessary?
I would say they make sense for a solution where the tooltip location is not directly connected to the mouse position but static on a specific position of the screen.

> Should the events elpNextTooltip and elpPrevTooltip also select the image and send out the imagefilename or not?
No, the filename should only be sent out, if a tooltip was selected. Maybe also see the documentation of the plugin included in this PR.

**Comment by deinhofer** [comment]
> No problem, thats what a review is for ;)
> The "problem" is that we need this plugin today for our evaluation with the client, but I will simply use an earlier commit.

uups, sorry. I was thinking about that possibility, but than I thought you would use the other PR #281.
Yes, please use an earlier commit:

`git checkout e93094085032c0b5e524ee7d8c6adb18831312f0`

should do it.


> 
> answering your remaining questions:
> 
> > Are the properties propInitX and propInitY really necessary?
> 
> I would say they make sense for a solution where the tooltip location is not directly connected to the mouse position but static on a specific position of the screen.
> 
> > Should the events elpNextTooltip and elpPrevTooltip also select the image and send out the imagefilename or not?
> 
> No, the filename should only be sent out, if a tooltip was selected. Maybe also see the documentation of the plugin included in this PR.

File: ARE/components/actuator.tooltip/src/main/java/eu/asterics/component/actuator/tooltip/GUI.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    }
+
+    /**
+     * Paints the Tooltip image at the defined location or at the mouse cursor location, if not set.
+     */
+    @Override
+    public void paint(Graphics g) {
+        // If the mouseX and mouseY values have been set from outside, use them, otherwise default to current mouse position.
+        if (mouseX < 0 || mouseY < 0) {
+            setLocation(MouseInfo.getPointerInfo().getLocation());
```
Comment (klues) [inline ARE/components/actuator.tooltip/src/main/java/eu/asterics/component/actuator/tooltip/GUI.java]: I think we should not use the mouse position here, but the `propInitialX` and `propInitialY` properties. Maybe `mouseX` and `mouseY` therefore is a bad name for the variables, the intention is/was that the tooltip plugin should be not natively related to the mouse position but only in connection with the Mouse plugin.

**Comment by klues** [comment]
@deinhofer: had to change the showTooltip() method and remove toggling of the window visibility because it breaks right and drag clicks of the mouse. Right-click-popups disappeared and drag-click was unintendetly released. see https://github.com/asterics/AsTeRICS/pull/283/commits/07daa3bbf526fc82c6aca22fc5eccdad3ab3a84a

**Comment by klues** [comment]
:star::star2::sparkles: :smiley: :sparkles::star2::star:

**Pull request merged**